File: MN/Practicando/Biseccion.py
''' METODO DE BISECCION PARA ECUACIONES NO LINEALES '''
import numpy as np

def biseccion(F,a,b,epsilon):
    f = lambda x: eval(F)
    c = np.array( [ (a+b)/2 ] , dtype = float )
    
    cant_iteraciones = int( np.ceil( (np.log( (b-a)/epsilon )) / ( np.log(2) ) ) )
    
    if f(a) * f(b) > 0:
        raise NameError('La funcion no cumple el teorema de Bolzano en el intervalo dado')

    for i in range(0,cant_iteraciones):
        
        if f(a) * f(b) == 0:
            if f(a) == 0:
                c = np.append( c,a )
                break
            elif f(b) == 0:
                c = np.append( c,b )
                break
        
        if f( c[i] ) == 0:
            break

        elif f(a) * f( c[i] ) < 0:
            b = c[i]
            c = np.append( c, (a+b)/2 )
        elif f( c[i] ) * f(b) < 0:
            a = c[i]
            c = np.append( c, (a+b)/2 )
    
    return c,f( c[i+1] ),cant_iteraciones

print( biseccion( 'x*np.sin(x)-1',0,2,1e-10 ) )

File: MN/Practicando/biseccion.m
f = inline('x*sin(x) - 1');

a = 0;
b = 2;
epsilon = 1e-5;
n = round( (log(b-a) - log(epsilon))/log(2) );

if f(a)*f(b)>0
	error('La funcion no cumple el teorema de Bolzano en el intervalo dado')
end

c = [ ];

for i = 1:1:n
	c(i) = (a+b)/2;
	if f(a)*f(c)<0
		b = c(i);
		c(i+1) = (a+b)/2;
		
	elseif f(c)*f(b)<0
		a = c(i);
		c(i+1) = (a+b)/2;
	
	end
end